prototype2/client/src/engine/control/event_distribution.rs

```diff
@@ -63,30 +72,48 @@ struct EventRouter<'a> {
 impl<'a> EventRouter<'a> {
   pub fn new(window: &'a glutin::Window,
              menu_state: &'a MenuState,
+             shift_held: &'a mut bool,
+             exit_flag: &'a mut ExitFlag,
              move_events: &'a mut Vec<MoveEvent>,
              menu_events: &'a mut Vec<MenuEvent>,
              camera_move_events: &'a mut Vec<CameraMoveEvent>)
              -> EventRouter<'a> {
     EventRouter {
       window: window,
       menu_state: menu_state,
+      shift_held: shift_held,
+      exit_flag: exit_flag,
       move_events: move_events,
       menu_events: menu_events,
       camera_move_events: camera_move_events,
     }
   }
 
   pub fn route_event(&mut self, event: glutin::Event) {
-    use glutin::VirtualKeyCode::{A, D, Escape, S, W};
+    use glutin::VirtualKeyCode::{A, D, Escape, LShift, RShift, S, W};
     use glutin::ElementState;
     match (self.menu_state, event) {
-      // Disabled, since menu was added. This will be moved into menu as a menu entry
-      // KeyboardInput(_, _, Some(Escape)) => *self.exit_flag = ExitFlag(true),
-      (&MenuState::Open, KeyboardInput(ElementState::Pressed, _, Some(Escape))) => {
-        self.menu_events.push(MenuEvent::Close)
+      (menu_state, KeyboardInput(ElementState::Pressed, _, Some(Escape))) => {
```

Comment:
sucks that `KeyboardInput` doesnt already have modifier keys baked in. Later might wanna look at building composite event w/ modifiers included